cachi2/core/package_managers/gomod.py

```diff
@@ -414,6 +428,88 @@ def _run(self, cmd: list[str], **kwargs: Any) -> str:
             ) from e
 
 
+class GoWork(UserDict):
+    """Representation of Go's go.work file."""
+
+    @staticmethod
+    def _get_go_work(go: Go, run_params: dict[str, Any]) -> str:
+        return go(["work", "edit", "-json"], run_params)
+
+    @staticmethod
+    def _get_go_work_path(app_dir: RootedPath) -> Optional[RootedPath]:
+        go_work_file = Go()(["env", "GOWORK"], {"cwd": app_dir}).rstrip()
+
+        # workspaces can be disabled explicitly with GOWORK=off
+        if not go_work_file or go_work_file == "off":
+            return None
+
+        # make sure that the path to go.work is within the request's root
+        return app_dir.join_within_root(go_work_file)
+
+    @property
+    def path(self) -> Optional[RootedPath]:
+        """Return the go.work file path."""
+        return self._path
+
+    @cached_property
+    def dir(self) -> Optional[RootedPath]:
+        """Return the base directory for the go.work file."""
+        if self._path is None:
+            return None
+
+        return RootedPath(self._app_dir.root).join_within_root(self._path.subpath_from_root.parent)
+
+    def __init__(self, app_dir: RootedPath) -> None:
+        """Initialize GoWork dict."""
+        super().__init__()
+        self._path = None
+        self._parsed = False
+        self._app_dir = app_dir
+
+        # workspaces may not be enabled -> empty instance
+        if (rooted_path := self._get_go_work_path(app_dir)) is None:
+            return
+
+        self._path = rooted_path
+
+    def __bool__(self) -> bool:
+        return self._path is not None and self.dir is not None
```

Comment:
`dir` looks to be `None` iff `_path` is `None`, is the second half of the expression needed?

Reply:
There were many occurrences of mypy complaining about `self.dir` being None. I think this tamed some of the complaints at some point during the development, so it's there on purpose. You're right it's absolutely not needed, but mypy doesn't know. Let me double check if mypy complains about something if I drop it again (might not be needed anymore, I had a ton of changes on top since I added this line).

Reply:
It feels like every other PR also has to wrangle mypy. If it is needed then it is needed, if it stays could you please add a short mypy-bashing comment?